pkg/client/cache/ttl_cache.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"time"
+)
+
+// TTLCache implements the store interface
+//	1. All entries are automatically time stamped on insert
+//		a. Thekey is computed based off the original item/keyFunc
+//		b. The value inserted under that key is the timestamped item
+//	2. TTL expiration happens lazily on read, via List/Get
+//	3. Time-stamps are stripped off unexpired entries before return
+type TTLCache struct {
+	cacheStorage     ThreadSafeStore
+	keyFunc          KeyFunc
+	expirationPolicy ExpirationPolicy
+}
+
+// ExpirationPolicy dictates when an object expires. Currently only abstracted out
+// so unittests don't rely on the system clock.
```

Comment:
I can see something like onExpire hooks in here, though there's no use case right now. Especially if we back out the threadSafeMap with somethign that talks to a db. The rename makes sense, I'll leave the rest in unless you prefer otherwise.